File: EmojiPanel/src/main/java/com/binlee/emoji/third/GlideImageEngine.java
package com.binlee.emoji.third;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.binlee.emoji.compat.ImageEngine;
import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.Rotate;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.bumptech.glide.request.FutureTarget;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.transition.Transition;

import java.io.File;
import java.util.concurrent.atomic.AtomicInteger;

public final class GlideImageEngine extends ImageEngine {

    private static final String TAG = "GlideImageEngine";

    @Override
    public void show(String url, ImageView target) {
        show(url, target, Config.asNull(), CALLBACK_NONE);
    }

    @Override
    public void showGif(String url, ImageView target) {
        show(url, target, Config.asGif(), CALLBACK_NONE);
    }

    @Override
    public void show(String url, ImageView target, Config config, Callback callback) {
        show(Drawable.class, url, target, config, callback);
    }

    @Override
    public <Res> void show(Class<Res> resClass, String url, ImageView target,
                           Config config, Callback callback) {
        final Config real = config == null ? Config.asNull() : config;
        Glide.with(target).as(resClass)
                .load(url)
                .apply(convertConfig(config))
                .listener(new RequestLogger<>())
                .into(new SimpleTarget<Res>() {
                    @Override
                    public void onResourceReady(@NonNull Res res,
                                                @Nullable Transition<? super Res> transition) {
                        if (res instanceof GifDrawable && real.mAsGif) {
                            target.setImageDrawable(((GifDrawable) res));
                            ((GifDrawable) res).start();
                        } else if (res instanceof Drawable) {
                            target.setImageDrawable(((Drawable) res));
                        } else if (res instanceof Bitmap) {
                            target.setImageBitmap(((Bitmap) res));
                        }
                    }
                });
    }

    @Override
    public void download(Context context, String url, Callback<File> callback) {
        if (callback == null) {
            throw new NullPointerException("download callback is null.");
        }
        final FutureTarget<File> target = Glide.with(context.getApplicationContext())
                .downloadOnly()
                .load(url)
                .listener(new RequestLogger<>())
                .submit();
        try {
            final File file = target.get();
            callback.onResReady(url, file);
        } catch (Throwable e) {
            callback.onFail(url, e);
            ANDROID.log(TAG, "#download error", e);
        }
    }

    @Override
    public void preload(Context context, PreloadCallback callback, String... urls) {
        if (urls == null || urls.length < 1) {
            throw new IllegalArgumentException("url is null or empty.");
        }
        final AtomicInteger counter = new AtomicInteger();
        final float unit = 100F / urls.length;
        final RequestManager mgr = Glide.with(context.getApplicationContext());
        final RequestLogger<Drawable> listener = new RequestLogger<Drawable>() {
            @Override
            public boolean onLoadFailed(@Nullable GlideException e, Object model,
                                        Target<Drawable> target, boolean isFirstResource) {
                onProgress(model, false);
                return false;
            }

            @Override
            public boolean onResourceReady(Drawable resource, Object model,
                                           Target<Drawable> target, DataSource dataSource,
                                           boolean isFirstResource) {
                onProgress(model, true);
                return false;
            }

            private void onProgress(Object model, boolean success) {
                if (callback == null) {
                    return;
                }
                final int count = counter.incrementAndGet();
                callback.onProgress(model, (int) (count * unit), success, count == urls.length);
            }
        };
        for (String url : urls) {
            mgr.load(url).listener(listener).preload();
        }
    }

    @Override
    protected RequestOptions convertConfig(Config config) {
        RequestOptions options = new RequestOptions();
        if (config == null) {
            return options.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC).dontTransform();
        }
        options = options.override(config.mWidth, config.mHeight);
        if (config.mCache == Cache.AUTOMATIC) {
            options = options.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC);
        } else if (config.mCache == Cache.RESOURCE) {
            options = options.diskCacheStrategy(DiskCacheStrategy.RESOURCE);
        } else if (config.mCache == Cache.DISK) {
            options = options.skipMemoryCache(true);
        } else if (config.mCache == Cache.NONE) {
            options = options.diskCacheStrategy(DiskCacheStrategy.NONE);
        } else {
            // maybe null
            options = options.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC);
        }
        if (config.mStyle == Style.CENTER_CROP) {
            options = options.centerCrop();
        } else if (config.mStyle == Style.CENTER_INSIDE) {
            options = options.centerInside();
        } else if (config.mStyle == Style.FIT_CENTER) {
            options = options.fitCenter();
        } else if (config.mStyle == Style.CIRCLE) {
            options = options.circleCrop();
        } else if (config.mStyle == Style.ROUNDED) {
            options = options.transform(new RoundedCorners(config.mRadius));
        } else if (config.mStyle == Style.ROTATE) {
            options = options.transform(new Rotate(config.mDegree));
        } else {
            options = options.dontTransform();
        }
        return options;
    }

    private static class RequestLogger<R> implements RequestListener<R> {

        private static final String DEF_TAG = "RequestLogger";
        private final String mTag;

        RequestLogger() {
            this(DEF_TAG);
        }

        RequestLogger(@NonNull String tag) {
            mTag = tag;
        }

        @Override
        public boolean onLoadFailed(@Nullable GlideException e, final Object model,
                                          Target<R> target, boolean isFirstResource) {
            ANDROID.log(mTag, "onLoadFailed() " + model, e);
            return false;
        }

        @Override
        public boolean onResourceReady(R resource, Object model, Target<R> target,
                                       DataSource dataSource, boolean isFirstResource) {
            ANDROID.log(mTag, "onResourceReady() " + model, null);
            return false;
        }
    }
}

File: EmojiPanel/src/main/java/com/binlee/emoji/third/GlideImageEngine.kt
package com.binlee.emoji.third

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.widget.ImageView
import com.binlee.emoji.compat.ImageEngine
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.Rotate
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.transition.Transition
import java.io.File
import java.util.concurrent.atomic.AtomicInteger

class GlideImageEngine : ImageEngine() {

    override fun show(url: String?, target: ImageView) {
        show(url, target, Config.asNull(), CALLBACK_NONE)
    }

    override fun showGif(url: String?, target: ImageView) {
        show(url, target, Config.asGif(), CALLBACK_NONE)
    }

    override fun show(url: String?, target: ImageView, config: Config?, callback: Callback<*>?) {
        show(Drawable::class.java, url, target, config, callback)
    }

    override fun <Res> show(resClass: Class<Res>?, url: String?, target: ImageView,
                            config: Config?, callback: Callback<*>?) {
        val real = config ?: Config.asNull()
        Glide.with(target).`as`(resClass!!)
                .load(url)
                .apply(convertConfig(config))
                .listener(RequestLogger())
                .into(object : SimpleTarget<Res>() {
                    override fun onResourceReady(res: Res,
                                                 transition: Transition<in Res>?) {
                        if (res is GifDrawable && real.mAsGif) {
                            target.setImageDrawable(res as GifDrawable)
                            (res as GifDrawable).start()
                        } else if (res is Drawable) {
                            target.setImageDrawable(res as Drawable)
                        } else if (res is Bitmap) {
                            target.setImageBitmap(res as Bitmap)
                        }
                    }
                })
    }

    override fun download(context: Context, url: String?, callback: Callback<File?>?) {
        if (callback == null) {
            throw NullPointerException("download callback is null.")
        }
        val target = Glide.with(context.applicationContext)
                .downloadOnly()
                .load(url)
                .listener(RequestLogger())
                .submit()
        try {
            val file = target.get()
            callback.onResReady(url, file)
        } catch (e: Throwable) {
            callback.onFail(url, e)
            ANDROID.log(TAG, "#download error", e)
        }
    }

    override fun preload(context: Context?, callback: PreloadCallback?, vararg urls: String?) {
        require(urls.isNotEmpty()) { "url is null or empty." }
        val counter = AtomicInteger()
        val unit = 100f / urls.size
        val mgr = Glide.with(context!!.applicationContext)
        val listener: RequestLogger<Drawable> = object : RequestLogger<Drawable>() {
            override fun onLoadFailed(e: GlideException?, model: Any,
                                      target: Target<Drawable>, isFirstResource: Boolean): Boolean {
                onProgress(model, false)
                return false
            }

            override fun onResourceReady(resource: Drawable, model: Any,
                                         target: Target<Drawable>, dataSource: DataSource,
                                         isFirstResource: Boolean): Boolean {
                onProgress(model, true)
                return false
            }

            private fun onProgress(model: Any, success: Boolean) {
                if (callback == null) {
                    return
                }
                val count = counter.incrementAndGet()
                callback.onProgress(model, (count * unit).toInt(), success, count == urls.size)
            }
        }
        for (url in urls) {
            mgr.load(url).listener(listener).preload()
        }
    }

    override fun convertConfig(config: Config?): RequestOptions {
        var options = RequestOptions()
        if (config == null) {
            return options.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC).dontTransform()
        }
        options = options.override(config.mWidth, config.mHeight)
        options = when (config.mCache) {
            Cache.AUTOMATIC -> {
                options.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
            }
            Cache.RESOURCE -> {
                options.diskCacheStrategy(DiskCacheStrategy.RESOURCE)
            }
            Cache.DISK -> {
                options.skipMemoryCache(true)
            }
            Cache.NONE -> {
                options.diskCacheStrategy(DiskCacheStrategy.NONE)
            }
            else -> {
                // maybe null
                options.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
            }
        }
        options = when (config.mStyle) {
            Style.CENTER_CROP -> {
                options.centerCrop()
            }
            Style.CENTER_INSIDE -> {
                options.centerInside()
            }
            Style.FIT_CENTER -> {
                options.fitCenter()
            }
            Style.CIRCLE -> {
                options.circleCrop()
            }
            Style.ROUNDED -> {
                options.transform(RoundedCorners(config.mRadius))
            }
            Style.ROTATE -> {
                options.transform(Rotate(config.mDegree))
            }
            else -> {
                options.dontTransform()
            }
        }
        return options
    }

    private open class RequestLogger<R> @JvmOverloads constructor(private val mTag: String = DEF_TAG) : RequestListener<R> {
        override fun onLoadFailed(e: GlideException?, model: Any,
                                  target: Target<R>, isFirstResource: Boolean): Boolean {
            ANDROID.log(mTag, "onLoadFailed() $model", e)
            return false
        }

        override fun onResourceReady(resource: R, model: Any, target: Target<R>,
                                     dataSource: DataSource, isFirstResource: Boolean): Boolean {
            ANDROID.log(mTag, "onResourceReady() $model", null)
            return false
        }

        companion object {
            private const val DEF_TAG = "RequestLogger"
        }
    }

    companion object {
        private const val TAG = "GlideImageEngine"
    }
}